### PR TITLE
Add S-LoRA

### PR DESCRIPTION
**Pull request by caoshiyi**

### CODE REVIEW

File: blog/2023-11-14-slora.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+previewImg: /images/blog/slora/serving_perf.png
+---
+In this blogpost, we introduce [S-LoRA](https://arxiv.org/abs/2311.03285) ([code](https://github.com/S-LoRA/S-LoRA)), a system designed for the scalable serving of many LoRA adapters. S-LoRA adopts the idea of
+
+1. **Unified Paging** for KV cache and adapter weights to reduce memory fragmentation. 
+2. **Heterogeneous Batching** of LoRA computation with different ranks leveraging optimized custom CUDA kernels which are aligned with the memory pool design.
+3. **S-LoRA TP** to ensure effective parallelization across multiple GPUs, incuring minimal communication cost for the added LoRA computation compared to that of the base model. 
+
+Evaluation results show that S-LoRA improves the throughput by up to 4 times and increase the number of served adapters by several orders of magnitude cpmpared to state-of-the-art libraries such as HuggingFace PEFT and vLLM (with naive support of LoRA serving).
+
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:14]: Put the result figure `/images/blog/slora/serving_perf.png` here

File: blog/2023-11-14-slora.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+\\
+&= xW + xAB.
+\end{align}
+$$
+
+### `x(W + AB)` v.s. `xW + xAB`
+
+One of the key innovations in the LoRA paper was the elimination of adapter inference latency by directly merging the adapter with the model parameters (as suggested by `Eq.(1)`). Additionally, to support multiple models on a single machine, the same paper proposes swapping adapters by adding and subtracting LoRA weights from the base model. While this approach enables low-latency inference for a single adapter and serial execution across adapters, it significantly reduces overall serving throughput and increases total latency when serving multiple adapters concurrently. We observe that the shared base model, which underpins numerous LoRA adapters, presents a substantial opportunity for batched inference. To achieve high-throughput multi-adapter serving, it is advantageous to separate the batchable base model computation from individual LoRA computations (as suggested by `Eq.(2)`).
+
+<img src="/images/blog/slora/lora_serving_plot_num_adapters.png" style="display:block; margin-top: auto; margin-left: auto; margin-right: auto; margin-bottom: auto; width: 50%"></img>
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:36]: Simplify this figure:
1. Remove alpha. Readers do not know what is "alpha".
2. Only keep the two curves with alpha = 1.

File: blog/2023-11-14-slora.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+### Model Settings
+
+| Setting | Base model | Hidden size | Adapter ranks   |
+| ------- | ---------- | ----------- | --------------- |
+| S1      | Llama-7B   | 4096        | {8}             |
+| S2      | Llama-7B   | 4096        | {64, 32, 16, 8} |
+| S4      | Llama-13B  | 5120        | {64, 32, 16}    |
+| S5      | Llama-30B  | 7168        | {32}            |
+| S6      | Llama-70B  | 8192        | {64}            |
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:89]: The results for other settings are not listed in the blog. Please paste the table from the paper here.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: blog/2023-11-14-slora.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+### Baselines
+
+We compare S-LoRA with HuggingFace PEFT and vLLM.
+
+1. PEFT stands for HuggingFace PEFT: We build a server using it that batches single adapter requests and switches adapter weights between batches.
+2. vLLM-packed: Since vLLM does not support LoRA, we merge the LoRA weights into the base model and serve the multiple versions of the merged weights separately. To serve m LoRA adapters, we run `m` vLLM workers on a single GPU, where multiple workers are separate processes managed by NVIDIA MPS.
+
+### Results
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:98]: Reorganize this section and try to add more results.

1. Comparison with PEFT and vLLM-packed. (serving_perf.png, Table. 3)
2. Tensor parallelism (Figure. 8)
3. One subfigure from Figure 5.

File: blog/2023-11-14-slora.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+h = xW' &= x(W+AB)
+\\
+&= xW + xAB.
+\end{align}
+$$
+
+### `x(W + AB)` v.s. `xW + xAB`
+
+One of the key innovations in the LoRA paper was the elimination of adapter inference latency by directly merging the adapter with the model parameters (as suggested by `Eq.(1)`). Additionally, to support multiple models on a single machine, the same paper proposes swapping adapters by adding and subtracting LoRA weights from the base model. While this approach enables low-latency inference for a single adapter and serial execution across adapters, it significantly reduces overall serving throughput and increases total latency when serving multiple adapters concurrently. We observe that the shared base model, which underpins numerous LoRA adapters, presents a substantial opportunity for batched inference. To achieve high-throughput multi-adapter serving, it is advantageous to separate the batchable base model computation from individual LoRA computations (as suggested by `Eq.(2)`).
+
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:35]: Put `public/images/blog/slora/batch_base_model.png` here, before the experiment results

File: blog/2023-11-14-slora.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+## Methods Summary
+
+1. **Unified Paging**: To reduce memory fragmentation and increase batch size, S-LoRA introduces a unified memory pool. This pool manages dynamic adapter weights and KV cache tensors by a unified paging mechanism.
+2. **Heterogeneous Batching**: To minimize the latency overhead when batching different adapters of varying ranks, S-LoRA employs highly optimized custom CUDA kernels. These kernels operate directly on non-contiguous memory and align with the memory pool design, facilitating efficient batched inference for LoRA.
+3. **S-LoRA TP**: To ensure effective parallelization across multiple GPUs, S-LoRA introduces a novel tensor parallelism strategy. This approach incurs minimal communication cost for the added LoRA computation compared to that of the base model. This is realized by scheduling communications on small intermediate tensors and fusing the large ones with the communications of the base model.
+
```
Comment (merrymercy) [inline blog/2023-11-14-slora.md:78]: Add the overview figure (Figure. 2) here

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Review by Ying1123 — APPROVED** [review]

**Pull request merged**